src/libexpr/eval-cache.cc

```diff
@@ -20,106 +20,172 @@ create table if not exists Attributes (
 
 struct AttrDb
 {
-    std::atomic_bool failed{false};
-
     const StoreDirConfig & cfg;
 
-    struct State
+    class State
     {
+        public:
         SQLite db;
-        SQLiteStmt insertAttribute;
-        SQLiteStmt insertAttributeWithContext;
-        SQLiteStmt queryAttribute;
-        SQLiteStmt queryAttributes;
-        std::unique_ptr<SQLiteTxn> txn;
+
+        private:
+        SQLiteStmt _upsertAttribute;
+        SQLiteStmt _insertAttribute;
+        SQLiteStmt _queryAttribute;
+        SQLiteStmt _queryAttributes;
```

Comment:
```suggestion
        SQLiteStmt upsertAttribute;
        SQLiteStmt insertAttribute;
        SQLiteStmt queryAttribute;
        SQLiteStmt queryAttributes;
```

Just to make it consistent with other classes.

Reply:
But then they conflict with methods of the same name. I introduced the methods to provide a clean api over sql queries (and to aggregate them, but it is no more useful now that we have upsert :-))